### PR TITLE
Protobufectomy: c# sdk

### PR DESCRIPTION
**Pull request by lcodes**
Depends on a change in the BSATN.Codegen library. This branch will be updated to reflect the changes to the csproj files once it's pushed.

## Description of Changes

- Removed JSON support (similar to changes being made on the TypeScript side)
- Removed all references to protobuf
- Swapped the protobuf codegen for the output of our own C# codegen in rust
- Refactored the connection class to reflect the API changes from protobuf to bsatn
- Updated the example to reflect these changes

## API

 - [ ] This is an API breaking change to the SDK

*If the API is breaking, please state below what will break*

## Requires SpacetimeDB PRs
*List any PRs here that are required for this SDK change to work*

### CODE REVIEW

File: src/SpacetimeDBClient.cs [hunk 1]
```diff
@@ -574,19 +587,19 @@ private void OnMessageProcessComplete(PreProcessedMessage preProcessed)
...
+                        clientIdentity = identityToken.Identity;
+                        var address = identityToken.Address; 
                         onIdentityReceived?.Invoke(identityToken.Token, clientIdentity, address);
                     }
                     catch (Exception e)
                     {
                         Logger.LogException(e);
                     }
                     break;
-                case { TypeCase: Message.TypeOneofCase.Event, Event: var event_ }:
```
Comment (jdetter) [inline src/SpacetimeDBClient.cs:589]: Why did we comment this out? BitCraft can't work without this.

**Review by jdetter — CHANGES_REQUESTED** [review]

File: src/SpacetimeDBClient.cs [hunk 2]
```diff
@@ -146,177 +145,192 @@ struct PreProcessedMessage
         private readonly CancellationTokenSource _preProcessCancellationTokenSource = new();
         private CancellationToken _preProcessCancellationToken => _preProcessCancellationTokenSource.Token;
 
+        struct PreProcessContext {
```
Comment (jdetter) [inline src/SpacetimeDBClient.cs:148]: What is the motivation behind this codeblock change - this code is *extremely* sensitive to changes. Over the last month I've probably spent over 20 hours just looking at the message processing code here and even the smallest change can completely break the SDK.

If we want to make these changes can we do it in a separate PR?

Comment (cloutiertyler) [reply src/SpacetimeDBClient.cs]: @lcodes This PR is to implement protobufecomy, not to refactor a bunch of the code. I would request that you revert the refactoring changes even if the code was working, although in this case it sounds like it's bugged.

If my understanding of the code is correct, then it is primarily refactoring. A change like this can in the best case do nothing and in the worst case introduce a bug right before the release of an Alpha playtest.

File: src/SpacetimeDBClient.cs [hunk 3]
```diff
@@ -201,122 +204,138 @@ HashSet<byte[]> GetInsertHashSet(System.Type tableType, int tableSize)
...
-                                {
-                                    table = table,
-                                    insert = new(obj, rowBytes),
-                                };
+                                        dbOps.Add(op);
+                                        break;
 
-                                dbOps.Add(op);
+                                    case EncodedValue.Text(var txt):
+                                        throw new InvalidOperationException("TODO?");
```
Comment (jdetter) [inline src/SpacetimeDBClient.cs:233]: In this case let's just print a warning and skip the row instead of throwing an exception.

**Review by jdetter — CHANGES_REQUESTED** [review]
Some requested changes for now but I'll take another look today if I have time

File: src/Utils.cs [hunk 4]
```diff
@@ -2,10 +2,10 @@
 using System.Collections.Generic;
 using System.Runtime.CompilerServices;
 
-#if UNITY_5_3_OR_NEWER
```
[the remark this answers is not in the document]
Comment (lcodes) [reply src/Utils.cs:5]: This attribute is relative to the .NET version, this change makes it work outside Unity as well.

(I remember hitting the compile error where IsExternalInit wasn't defined, and copied this snipped from a past project of mine.)

File: src/WebSocket.cs [hunk 5]
```diff
@@ -183,7 +185,10 @@ private async Task ProcessSendQueue()
                         }
                     }
 
-                    await Ws!.SendAsync(message, WebSocketMessageType.Binary, true, CancellationToken.None);
+                    var o = new MemoryStream();
+                    var w = new BinaryWriter(o);
```
[the remark this answers is not in the document]
Comment (lcodes) [reply src/WebSocket.cs:189]: Yeah every message is overkill.

I'd suggest taking note and fixing perf issues in another PR

File: src/SpacetimeDBClient.cs [hunk 6]
```diff
@@ -146,29 +145,32 @@ struct PreProcessedMessage
...
+        };
+
         void PreProcessMessages()
         {
-            while (!isClosing)
-            {
-                try
-                {
+            try {
+                while (!isClosing) {
```
[the remark this answers is not in the document]
Comment (lcodes) [reply src/SpacetimeDBClient.cs:157]: Same behavior, but now creates only one exception frame for the whole function instead of once per loop iteration.

File: src/SpacetimeDBClient.cs (hunk 3 above)
Comment (lcodes) [reply src/SpacetimeDBClient.cs:233]: Good point.

And besides, the SDK should only communicate in BSATN with the server now.

File: src/SpacetimeDBClient.cs [hunk 7]
```diff
@@ -201,122 +204,138 @@ HashSet<byte[]> GetInsertHashSet(System.Type tableType, int tableSize)
...
+                                try {
+                                    transactionUpdate.ReducerCall.ReducerEvent = ReducerEventFromDbEvent(transactionUpdate);
+                                }
+                                catch (Exception e) {
+                                    Logger.LogException(e);
+                                }
+                                break;
+                            case UpdateStatus.Failed(var failed):
+                                break;
+                            case UpdateStatus.OutOfEnergy(var outOfEnergy):
```
[the remark this answers is not in the document]
Comment (lcodes) [reply src/SpacetimeDBClient.cs:328]: Good idea.

File: src/SpacetimeDBClient.cs [hunk 8]
```diff
@@ -80,7 +78,8 @@ struct DbOp
         /// <summary>
         /// Invoked when an event message is received or at the end of a transaction update.
         /// </summary>
-        public event Action<ClientApi.Event>? onEvent;
+        public event Action<TransactionUpdate>? onEvent;
+        public event Action<OneOffQueryResponse>? onResult;
```
[the remark this answers is not in the document]
Comment (lcodes) [reply src/SpacetimeDBClient.cs:82]: I have mixed feelings about it. Because it's a breaking change either way.

The reason it's TransactionUpdate now is because it's the closest match to the previous Event. But then passing ServerMessage to the SDK users means they get to dispatch on its enum variant.

If that's fine, happy to make the change.

File: src/SpacetimeDBClient.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (lcodes) [reply src/SpacetimeDBClient.cs:157]: Yeah it's a good example of a refactor that could've been another commit or PR :)

File: src/SpacetimeDBClient.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (jdetter) [reply src/SpacetimeDBClient.cs:82]: Let's go with this:

```C#
public event Action<ServerMessage>? onEvent;
```

**Review by jdetter — APPROVED** [review]
Tested against BitCraft in the Unity SDK, appears to be working just fine. Thanks @lcodes!

File: src/SpacetimeDBClient.cs [hunk 9]
```diff
@@ -146,29 +145,38 @@ struct PreProcessedMessage
...
+            EncodedValue.Text(var text) => throw new InvalidOperationException("JavaScript messages aren't supported."),
+            _ => throw new InvalidOperationException(),
+        };
+
         void PreProcessMessages()
         {
-            while (!isClosing)
+            try
             {
-                try
```
Comment (jdetter) [inline src/SpacetimeDBClient.cs:153]: Why did we swap the ordering of this? It was intentional that we had the try on the inside.
